src/rules/unnecessaryElseRule.ts

```diff
@@ -0,0 +1,100 @@
+/**
+ * @license
+ * Copyright 2019 Palantir Technologies, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import * as utils from "tsutils";
+import * as ts from "typescript";
+
+import * as Lint from "../index";
+
+import { codeExamples } from "./code-examples/unnecessaryElse.examples";
+
+export class Rule extends Lint.Rules.AbstractRule {
+    /* tslint:disable:object-literal-sort-keys */
+    public static metadata: Lint.IRuleMetadata = {
+        description: Lint.Utils.dedent`
+        Disallows \`else\` blocks following \`if\` blocks ending with a \`break\`, \`continue\`, \`return\`, or \`throw\` statement.`,
+        descriptionDetails: "",
+        optionExamples: [true],
+        options: null,
+        optionsDescription: "Not configurable.",
+        rationale: Lint.Utils.dedent`
+        When an \`if\` block is guaranteed to exit control flow when entered,
+        it is unnecessary to add an \`else\` statement.
+        The contents that would be in the \`else\` block can be placed after the end of the \`if\` block.`,
+        ruleName: "unnecessary-else",
+        type: "functionality",
+        typescriptOnly: false,
+        codeExamples,
+    };
+    /* tslint:disable:object-literal-sort-keys */
+
+    public static FAILURE_STRING(name: string): string {
+        return `The preceding \`if\` block ends with a \`${name}\` statement. This \`else\` block is unnecessary.`;
+    }
+
+    public apply(sourceFile: ts.SourceFile): Lint.RuleFailure[] {
+        return this.applyWithFunction(sourceFile, walk);
+    }
+}
+
+function walk(ctx: Lint.WalkContext<void>): void {
+    ts.forEachChild(ctx.sourceFile, function cb(node: ts.Node): void {
+        if (utils.isIfStatement(node)) {
+            const jumpStatement = getJumpStatement(
+                getLastStatement(node.thenStatement as ts.Block),
+            );
+            if (jumpStatement !== undefined && node.elseStatement !== undefined) {
+                ctx.addFailureAtNode(
+                    node.elseStatement.getChildAt(0),
+                    Rule.FAILURE_STRING(jumpStatement),
+                );
+                ts.forEachChild(node.elseStatement, cb);
+            }
+            ts.forEachChild(node.thenStatement, cb);
+        } else {
+            return ts.forEachChild(node, cb);
```

Comment:
It looks like you have a few different `ts.forEachChild`s: individual ones for the node's `elseStatement` and `thenStatement` if it's an `IfStatement`, but just a regular recursion otherwise? You should be able to simplify this a bit by only having one call to `ts.forEachChild` that's always called.

Also, if `jumpStatement` is undefined but `node.elseStatement` isn't, this code will skip it. That seems like a potential bug?

Reply:
Done https://github.com/palantir/tslint/pull/4502/commits/be2f8b4323b7b534130489a8467a91b6962d745e

Reply:
Sorry @debsmita1 I think I wasn't very clear here. I'm suggesting you don't need to have `ts.forEachChild` statements in the big `if (utils.isIfStatement` block:

```ts
if (utils.isIfStatement(node)) {
    // all your glorious rule logic here
    // no ts.forEachChild in this area
}

ts.forEachChild(node, cb);
```

Note the lack of `else` statement: a single `ts.forEachChild` should recurse on all the node's children no matter what. Could you please switch to that? It's a bit less code and better practice to minimize recursive calls (since they can get tricky in larger rules).

Reply:
Done https://github.com/palantir/tslint/pull/4502/commits/f36d4201e2fa8ec3aec60a022db2bf51435aceaf